src/dtypes/amp.rs

```diff
@@ -0,0 +1,571 @@
+where
+    &'l F: std::ops::Add<&'r F, Output = F>,
+{
+    type Output = AMP<F>;
+    fn add(self, rhs: &'r AMP<F>) -> Self::Output {
+        AMP(&self.0 + &rhs.0)
+    }
+}
+
+impl<'l, 'r, F> std::ops::Sub<&'r AMP<F>> for &'l AMP<F>
+where
+    &'l F: std::ops::Sub<&'r F, Output = F>,
+{
+    type Output = AMP<F>;
+    fn sub(self, rhs: &'r AMP<F>) -> Self::Output {
+        AMP(&self.0 - &rhs.0)
+    }
+}
+
+impl<'l, 'r, F> std::ops::Mul<&'r AMP<F>> for &'l AMP<F>
+where
+    &'l F: std::ops::Mul<&'r F, Output = F>,
+{
+    type Output = AMP<F>;
+    fn mul(self, rhs: &'r AMP<F>) -> Self::Output {
+        AMP(&self.0 * &rhs.0)
+    }
+}
+
+impl<'l, 'r, F> std::ops::Div<&'r AMP<F>> for &'l AMP<F>
+where
+    &'l F: std::ops::Div<&'r F, Output = F>,
+{
+    type Output = AMP<F>;
+    fn div(self, rhs: &'r AMP<F>) -> Self::Output {
+        AMP(&self.0 / &rhs.0)
+    }
+}
+
+impl<'l, 'r, F> std::ops::Rem<&'r AMP<F>> for &'l AMP<F>
+where
+    &'l F: std::ops::Rem<&'r F, Output = F>,
+{
+    type Output = AMP<F>;
+    fn rem(self, rhs: &'r AMP<F>) -> Self::Output {
+        AMP(&self.0 % &rhs.0)
+    }
+}
+
+impl<'l, F> std::ops::Neg for &'l AMP<F>
+where
+    &'l F: std::ops::Neg<Output = F>,
+{
+    type Output = AMP<F>;
+    fn neg(self) -> Self::Output {
+        AMP(-&self.0)
+    }
+}
+
+impl<F: std::ops::AddAssign<F>> std::ops::AddAssign<AMP<F>> for AMP<F> {
+    fn add_assign(&mut self, rhs: Self) {
+        self.0 += rhs.0;
+    }
+}
+
+impl<F: std::ops::SubAssign<F>> std::ops::SubAssign<AMP<F>> for AMP<F> {
+    fn sub_assign(&mut self, rhs: Self) {
+        self.0 -= rhs.0;
+    }
+}
+
+impl<F: std::ops::MulAssign<F>> std::ops::MulAssign<AMP<F>> for AMP<F> {
+    fn mul_assign(&mut self, rhs: Self) {
+        self.0 *= rhs.0;
+    }
+}
+
+impl<F: std::ops::DivAssign<F>> std::ops::DivAssign<AMP<F>> for AMP<F> {
+    fn div_assign(&mut self, rhs: Self) {
+        self.0 /= rhs.0;
+    }
+}
+
+impl<F: num_traits::FromPrimitive> num_traits::FromPrimitive for AMP<F> {
+    fn from_f32(n: f32) -> Option<Self> {
+        F::from_f32(n).map(AMP)
+    }
+    fn from_f64(n: f64) -> Option<Self> {
+        F::from_f64(n).map(AMP)
+    }
+    fn from_i64(n: i64) -> Option<Self> {
+        F::from_i64(n).map(AMP)
+    }
+    fn from_u64(n: u64) -> Option<Self> {
+        F::from_u64(n).map(AMP)
+    }
+}
+
+impl<T: 'static + Copy, F: num_traits::AsPrimitive<T>> num_traits::AsPrimitive<T> for AMP<F> {
+    fn as_(self) -> T {
+        self.0.as_()
+    }
+}
+
+#[cfg(feature = "f16")]
+impl num_traits::AsPrimitive<AMP<half::f16>> for half::f16 {
+    fn as_(self) -> AMP<half::f16> {
+        AMP(self)
+    }
+}
+
+#[cfg(feature = "f16")]
+impl num_traits::AsPrimitive<AMP<half::f16>> for f32 {
+    fn as_(self) -> AMP<half::f16> {
+        AMP(half::f16::from_f32(self))
+    }
+}
+
+#[cfg(feature = "f16")]
+impl num_traits::AsPrimitive<AMP<half::f16>> for f64 {
+    fn as_(self) -> AMP<half::f16> {
+        AMP(half::f16::from_f64(self))
+    }
+}
+
```

Comment:
```suggestion
#[cfg(feature = "f16")]
impl<T: num_traits::AsPrimitive<E>, E> num_traits::AsPrimitive<AMP<E>> for T {
    fn as_(self) -> AMP<E> {
        AMP(self.as_())
    }
}
```

Reply:
Actually this leads to unconstrained type error on T